### PR TITLE
Remove predicate pushdown switch in TPCH connector (expose and test for #9800 bug)

### PR DESCRIPTION
**Pull request by kokosing**

### CODE REVIEW

**Comment by kokosing** [comment]
if you apply this PR and revert 3eac47dfec then you will get an error mentioned in #9800

**Comment by sopel39** [comment]
@findepi yesterday discovered a bug in TPCH Predicate pushdown code that caused layouts to be completely filtered when using column other than `ordestatus`. I think that fix should be part of this PR or land earlier.

File: presto-main/src/test/java/com/facebook/presto/cost/TestCoefficientBasedStatsCalculator.java [hunk 1]
```diff
@@ -45,7 +45,7 @@ public TestCoefficientBasedStatsCalculator()
 
         queryRunner.createCatalog(
                 queryRunner.getDefaultSession().getCatalog().get(),
-                new TpchConnectorFactory(1, true),
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/cost/TestCoefficientBasedStatsCalculator.java]: > In regards of testing for case with predicate pushdown enabled
and disabled, one can select columns which either support or do not
support predicate according to test needs.

I kind of disagree. The set of columns that have support for predicate push down is allowed to grow and it will grow, e.g. when someone writes a test for push down of predicates on two columns.

Comment (kokosing) [reply presto-main/src/test/java/com/facebook/presto/cost/TestCoefficientBasedStatsCalculator.java]: It is true, but in regards of TPCH the set of columns where you could pushdown the predicate is limited to dictionary like columns.

**Comment by kokosing** [comment]
@sopel39 @findepi Rebased. May I ask for a review.

File: presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java [hunk 2]
```diff
@@ -200,32 +208,35 @@ public void close()
 
         private boolean rowMatchesPredicate()
         {
-            if (!predicate.isPresent()) {
+            if (!predicate.isPresent() || predicate.get().isAll()) {
                 return true;
             }
-            return predicate.get().test(rowMap());
-        }
+            Set<ColumnHandle> predicatePushdownColumns = predicate.get().getDomains().get().keySet();
```
Comment (sopel39) [inline presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java]: why not:
```
Set<TpchColumnHandle> predicatePushdownColumns = predicate.get().getDomains().get().keySet().stream()
    .map(column -> (TpchColumnHandle) column)
    .collect(toImmutableSet());
```

File: presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java [hunk 3]
```diff
@@ -21,7 +21,6 @@
 import com.facebook.presto.spi.predicate.TupleDomain;
 import com.facebook.presto.spi.type.Type;
 import com.google.common.collect.ImmutableList;
-import com.google.common.collect.ImmutableMap;
```
Comment (sopel39) [inline presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java]: commit message is not clear.

File: presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java [hunk 4]
```diff
@@ -200,32 +208,35 @@ public void close()
...
-        {
-            ImmutableMap.Builder<ColumnHandle, NullableValue> builder = ImmutableMap.builder();
-            for (int field = 0; field < columnHandles.size(); ++field) {
-                Type type = columnTypes.get(field);
-                builder.put(columnHandles.get(field), NullableValue.of(type, getPrestoObject(field, type)));
-            }
-            return builder.build();
+            TupleDomain<ColumnHandle> rowTupleDomain = TupleDomain.fromFixedValues(rowMap);
+
+            return predicate.get().contains(rowTupleDomain);
```
Comment (sopel39) [inline presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java]: does it work when `predicate` does not contain all columns from row (I think in that case it will always return false)?

File: presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java [hunk 5]
```diff
@@ -200,32 +208,35 @@ public void close()
 
         private boolean rowMatchesPredicate()
         {
-            if (!predicate.isPresent()) {
+            if (!predicate.isPresent() || predicate.get().isAll()) {
```
Comment (findepi) [inline presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java]: You could remove `Optional`-ness of the predicate in the constructor and handle just single case here.

File: presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java [hunk 6]
```diff
@@ -200,32 +208,35 @@ public void close()
...
+            Set<ColumnHandle> predicatePushdownColumns = predicate.get().getDomains().get().keySet();
+            Map<ColumnHandle, NullableValue> rowMap = predicatePushdownColumns.stream()
+                    .collect(toImmutableMap(
+                            column -> column,
+                            column -> {
+                                TpchColumnHandle tpchColumnHandle = (TpchColumnHandle) column;
+                                Type type = tpchColumnHandle.getType();
+                                TpchColumn tpchColumn = table.getColumn(tpchColumnHandle.getColumnName());
+                                return NullableValue.of(type, getPrestoObject(tpchColumn, type));
+                            }));
```
Comment (findepi) [inline presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java]: I don't understand. Does this need to be so complex? Would constraint evaluator from https://github.com/prestodb/presto/pull/9871 help?

File: presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java [hunk 7]
```diff
@@ -200,32 +208,35 @@ public void close()
...
-                builder.put(columnHandles.get(field), NullableValue.of(type, getPrestoObject(field, type)));
-            }
-            return builder.build();
+            TupleDomain<ColumnHandle> rowTupleDomain = TupleDomain.fromFixedValues(rowMap);
+
+            return predicate.get().contains(rowTupleDomain);
         }
 
-        private Object getPrestoObject(int field, Type type)
+        private Object getPrestoObject(TpchColumn<E> column, Type type)
```
Comment (findepi) [inline presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java]: Maybe there should be `TpchColumn::getObject` method: `column.getObject(row)` ?

File: presto-tpch/src/test/java/com/facebook/presto/tpch/TestTpchMetadata.java [hunk 8]
```diff
@@ -359,4 +360,9 @@ private ColumnStatistics createColumnStatistics(Optional<Double> distinctValuesC
                 .setNullsFraction(zeroValue())
                 .build();
     }
+
+    private static Predicate<Map<ColumnHandle, NullableValue>> convertToPredicate(TupleDomain<ColumnHandle> tupleDomain)
+    {
+        return bindings -> tupleDomain.contains(TupleDomain.fromFixedValues(bindings));
```
Comment (findepi) [inline presto-tpch/src/test/java/com/facebook/presto/tpch/TestTpchMetadata.java]: Are you sure? `TupleDomain::contains` caused the problem in #9839, see https://github.com/prestodb/presto/pull/9839/commits/d4a0f2d9f03f1024449a5319a87edef549286616#diff-ccd5d11614e67178cf1dfe85be861160L411

File: presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java (hunk 2 above)
Comment (findepi) [inline presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java]: inline?

File: presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java [hunk 9]
```diff
@@ -200,32 +208,35 @@ public void close()
...
-            return predicate.get().test(rowMap());
-        }
+            Set<ColumnHandle> predicatePushdownColumns = predicate.get().getDomains().get().keySet();
+            Map<ColumnHandle, NullableValue> rowMap = predicatePushdownColumns.stream()
+                    .collect(toImmutableMap(
+                            column -> column,
+                            column -> {
+                                TpchColumnHandle tpchColumnHandle = (TpchColumnHandle) column;
+                                Type type = tpchColumnHandle.getType();
+                                TpchColumn tpchColumn = table.getColumn(tpchColumnHandle.getColumnName());
```
Comment (findepi) [inline presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java]: Does this work regardless of `tpch.column-naming`?

File: presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java (hunk 4 above)
Comment (findepi) [inline presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java]: Previously the `.test()` method was used. Are you sure `.contains()` is correct? `TupleDomain::contains` caused the problem in #9839, see https://github.com/prestodb/presto/pull/9839/commits/d4a0f2d9f03f1024449a5319a87edef549286616#diff-ccd5d11614e67178cf1dfe85be861160L411

File: presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java [hunk 10]
```diff
@@ -67,13 +67,15 @@
     private final List<Type> columnTypes;
     private final Optional<TupleDomain<ColumnHandle>> predicate;
 
-    public TpchRecordSet(Iterable<E> rows, TpchTable<E> table, Iterable<TpchColumn<E>> columns, Optional<TupleDomain<ColumnHandle>> predicate)
```
Comment (findepi) [inline presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java]: Why this change?

File: presto-main/src/test/java/com/facebook/presto/cost/TestCoefficientBasedStatsCalculator.java (hunk 1 above)
Comment (findepi) [reply presto-main/src/test/java/com/facebook/presto/cost/TestCoefficientBasedStatsCalculator.java]: What about changing the default, instead of removing the switch?

File: presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java (hunk 9 above)
Comment (kokosing) [reply presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java]: Yes.

File: presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java (hunk 6 above)
Comment (kokosing) [reply presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java]: Unfortunately it looks like it has to be so "complex". Constraint evaluator is much more complex, it works for expressions and is able to call functions etc.

File: presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java (hunk 4 above)
Comment (kokosing) [reply presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java]: Yes. Because `rowTupleDomain` contains only columns which are in predicate. It is build from `predicate` columns. See how `rowMap` is built.

Comment (kokosing) [reply presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java]: Yes. See the response for https://github.com/prestodb/presto/pull/9823/commits/bda705b783589b31fd86c20283c0782ade6f7164#r165374015

File: presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java (hunk 7 above)
Comment (kokosing) [reply presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java]: `TpchColumn` is a `tpch` library class, it does not know anything about Presto objects.

File: presto-tpch/src/test/java/com/facebook/presto/tpch/TestTpchMetadata.java (hunk 8 above)
Comment (kokosing) [reply presto-tpch/src/test/java/com/facebook/presto/tpch/TestTpchMetadata.java]: Here, I am not 100% sure. But this is the method you used in #9839. I just moved it here from `TpchResultSet`.

File: presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java (hunk 10 above)
Comment (kokosing) [reply presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java]: Because I didn't like: `ImmutableList.copyOf(transform(columns, TpchMetadata::getPrestoType))` below. And because of that nobody really cared that is `Iterable`, so there is no need to complicate things.

File: presto-main/src/test/java/com/facebook/presto/cost/TestCoefficientBasedStatsCalculator.java (hunk 1 above)
Comment (kokosing) [reply presto-main/src/test/java/com/facebook/presto/cost/TestCoefficientBasedStatsCalculator.java]: `There is no practical reason that one could prefer predicate pushdown
disabled.`

Switch makes things more complicated and here I do not see added value, but I only see a cost.

Comment (kokosing) [reply presto-main/src/test/java/com/facebook/presto/cost/TestCoefficientBasedStatsCalculator.java]: `There is no practical reason that one could prefer predicate pushdown
disabled.`

Switch makes things more complicated and here I do not see added value, but I only see a cost.

For example because of this switch, predicate pushdown practically was wrongly implemented for about year and nobody noticed that.

File: presto-tpch/src/test/java/com/facebook/presto/tpch/TestTpchMetadata.java [hunk 11]
```diff
@@ -359,4 +360,9 @@ private ColumnStatistics createColumnStatistics(Optional<Double> distinctValuesC
                 .setNullsFraction(zeroValue())
                 .build();
     }
+
+    private static Predicate<Map<ColumnHandle, NullableValue>> convertToPredicate(TupleDomain<ColumnHandle> tupleDomain)
```
Comment (sopel39) [inline presto-tpch/src/test/java/com/facebook/presto/tpch/TestTpchMetadata.java]: this is unused now

File: presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java [hunk 12]
```diff
@@ -197,26 +204,36 @@ public void close()
...
+                    .collect(toImmutableMap(
+                            column -> column,
+                            column -> {
+                                TpchColumnHandle tpchColumnHandle = (TpchColumnHandle) column;
+                                Type type = tpchColumnHandle.getType();
+                                TpchColumn tpchColumn = table.getColumn(tpchColumnHandle.getColumnName());
+                                return NullableValue.of(type, getPrestoObject(tpchColumn, type));
+                            }));
+
+            TupleDomain<ColumnHandle> rowTupleDomain = TupleDomain.fromFixedValues(rowMap);
```
Comment (sopel39) [inline presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java]: inline

File: presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java [hunk 13]
```diff
@@ -198,14 +197,6 @@ public void close()
             closed = true;
         }
 
-        private boolean rowMatchesPredicate()
```
Comment (sopel39) [inline presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java]: you remove this method only to add it later on? Are the commits ordered properly?

File: presto-tpch/src/test/java/com/facebook/presto/tpch/TestTpchMetadata.java (hunk 8 above)
Comment (findepi) [reply presto-tpch/src/test/java/com/facebook/presto/tpch/TestTpchMetadata.java]: But there, there was also `tupleDomain = filterColumns(tupleDomain, ...)`

File: presto-main/src/test/java/com/facebook/presto/cost/TestCoefficientBasedStatsCalculator.java (hunk 1 above)
Comment (findepi) [reply presto-main/src/test/java/com/facebook/presto/cost/TestCoefficientBasedStatsCalculator.java]: Yea, but what about testing how Presto behaves when a connector doesn't have predicate push down? I am worried that we're losing test coverage here. Of course, we're also gaining coverage of tpch connector itself, and also for how Presto behaves when a connector _does_ have predicate push down... This is why I propose to retain the switch.

File: presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java [hunk 14]
```diff
@@ -197,26 +204,36 @@ public void close()
...
+                            column -> {
+                                TpchColumnHandle tpchColumnHandle = (TpchColumnHandle) column;
+                                Type type = tpchColumnHandle.getType();
+                                TpchColumn tpchColumn = table.getColumn(tpchColumnHandle.getColumnName());
+                                return NullableValue.of(type, getPrestoObject(tpchColumn, type));
+                            }));
+
+            TupleDomain<ColumnHandle> rowTupleDomain = TupleDomain.fromFixedValues(rowMap);
+
+            return predicate.contains(rowTupleDomain);
```
Comment (sopel39) [inline presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java]: I think this will be always `false` when `predicate` doesn't contain all table columns. Please validate.

File: presto-main/src/test/java/com/facebook/presto/cost/TestCoefficientBasedStatsCalculator.java (hunk 1 above)
Comment (kokosing) [reply presto-main/src/test/java/com/facebook/presto/cost/TestCoefficientBasedStatsCalculator.java]: > In regards of testing for case with predicate pushdown enabled
and disabled, one can select columns which either support or do not
support predicate according to test needs.

TPCH supports predicate pushdown for two tables. So this is a kind of hybrid connector which can be used for both kind of tests.

File: presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java (hunk 14 above)
Comment (sopel39) [reply presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java]: Ok if you say so.

However, `TupleDomain#contains` does `columnWiseUnion` which shouldn't be equal to `predicate` if `rowTupleDomain` contains more columns than `predicate`. I remember @findepi hit the same issue

Maybe add test to `AbstractTestQueries`:
```
select * from orders where orderstatus='F'
```
?

Comment (kokosing) [reply presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java]: > if rowTupleDomain contains more columns than predicate

`rowTupleDomain` will have exactly the same number of columns as `predicate`, as it is build from `predicate`.

`AbstractTestQueries` already contains queries containing: `SELECT .. FROM orders WHERE orderstatus = 'F' `
